### ISSUE TITLE
Handle weighted samples

### ISSUE THREAD

**Issue by devmotion** [TuringLang/MCMCChains.jl#253]
Currently, MCMCChains does not handle weighted samples such as the ones output by IS correctly and reports meaningless statistics. This came up in https://github.com/TuringLang/Turing.jl/issues/1467.

Maybe one approach could be to add an additional field `weights`, which by default would a singleton such as `UniformWeights` but would allow to specify `LogWeights` (or more generally any weight from StatsBase?) as well.

**Comment by cpfiffer** [comment]
I think we used to have this in the super old `MCMCChains` but it was removed by assuming that it would be added as a parameter if it was important. Is there any reason we couldn't just do that?

**Comment by devmotion** [comment]
You mean by saving it in the AxisArray? Similar to all the Turing statistics and other additional parameters, this could lead to name conflicts with "actual" parameters but in this case with much more severe (and possibly unintended) consequences.

**Comment by devmotion** [comment]
Actually, I think that one could even consider separating the actual parameters and all the metadata/statistics into different arrays to avoid these problems in the current implementation.

**Comment by cpfiffer** [comment]
Yeah, that would actually be quite nice -- tracking them separately would be a lot better than this weirdo group map thing we're doing internally.

**Comment by ParadaCarleton** [comment]
This is probably a lot more important to do now with ParetoSmooth.jl; we'll need to deal with weighted samples if we want to let people use PSIS as an inference algorithm rather than just for model comparison. Maybe some of the work already in ParetoSmooth can be used for this? If weights aren't working, I assume SMC diagnostics aren't working either?

**Comment by mileslucas** [comment]
[NestedSamplers.jl uses weights](https://github.com/TuringLang/NestedSamplers.jl/blob/c5aa3e947ff74feddf2ac6a94b9243a008422223/src/step.jl#L134) and could benefit from this functionality being built in. I think using the StatsBase weights is a good idea, too.

**Comment by mschauer** [comment]
Our use case would be that https://github.com/rubenseyer/FFFSampler.jl spends a random time in each point which constitutes a weighted sample, a fact we need to hide by discretizing the trace to a grid right now.
